aten/src/ATen/DeviceGuard.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include <ATen/Device.h>
+#include <ATen/ScalarType.h>
+#include <ATen/Tensor.h>
+#include <ATen/detail/CUDAHooksInterface.h>
+#include <ATen/optional.h>
+
+#include <cstddef>
+
+namespace at {
+/// RAII guard that sets a certain default GPU index in its constructor, and
+/// changes it back to the device that was originally active upon destruction.
+///
+/// The index is always reset to the one that was active at the time of
```

Comment:
Is this standard with RAII guards?  I wouldn't know if the reset behavior was the previous or first device, which means I have to read the documentation carefully.  But maybe this is standard and I don't know.

Also, if it's standard, is it particularly useful?  Can you point to an example?

Reply:
This was the previous behavior of `AutoGPU`. I also think that instantiation of a `DeviceGuard` should be like an anchor/checkpoint, and whatever you do you get back to this checkpoint.

Reply:
That sounds fine, it just always seemed a little strange to me that we allow `set_index` instead of just using another guard.